tests/test_param_routing.py

```diff
@@ -0,0 +1,115 @@
+import inspect
+
+from distutils.version import LooseVersion
+from typing import Any
+from typing import Dict
+
+import numpy as np
+import pytest
+import tensorflow as tf
+
+from tensorflow.keras import Model
+
+from scikeras.wrappers import BaseWrapper
+from scikeras.wrappers import KerasClassifier
+from scikeras.wrappers import KerasRegressor
+
+from .mlp_models import dynamic_classifier
+from .mlp_models import dynamic_regressor
+
+
+def test_routing_basic():
+    """Tests that parameters are routed to the correct place based on static
+    dictionaries.
+    """
+    n, d = 20, 3
+    n_classes = 3
+    X = np.random.uniform(size=(n, d)).astype(float)
+    y = np.random.choice(n_classes, size=n).astype(int)
+
+    def build_fn(*args, **kwargs):
+        assert len(args) == 0, "No *args should be passed to `build_fn`"
+        assert tuple(kwargs.keys()) == (
+            "hidden_layer_sizes",
+            "meta_params",
+            "compile_params",
+        ), "The number and order of **kwargs passed to `build_fn` should be fixed"
+        meta = set(kwargs["meta_params"].keys()) - {"X", "y"}
+        expected_meta = KerasClassifier._meta_params - {
+            "model_",
+            "history_",
+            "is_fitted_",
+        }
+        assert meta == expected_meta
```

Comment:
I want `KerasClassifier._meta_params` to be the source of truth as to what meta parameters exist. That way the users can check them, and we can have a test (this one) that validates that. In the future I'd also like to have a docs test that checks that each one of them is documented.

Reply:
> keras_exected_n_outputs



> How are `n_outputs`, `keras_exected_n_outputs` determined?

That's is going to have to be part of the documentation for the attributes of each wrapper, i.e. `_meta_params`.

Reply:
I'm not a huge fan of the names `n_outputs_` and `keras_expected_n_outputs`: I don't know the difference or how they're computed.

I've looked at how they're computed. Maybe do the following renamings?

* `n_outputs_` → `num_outputs__data` or `y.shape[1]`
* `keras_expected_n_outputs` → `num_outputs__expected`
* `cls_type_` → `target_type`
* `input_dtype_` → `X.dtype`

We're already breaking the API; we can always include the other keys but deprecate them (I think Sklearn has a utility just for that).

> That way the users can check [`_meta_params`]

Then why does `_meta_params` have a leading underscore? Doesn't that mean that attribute is private in Python parlance? Why does a certain key, `_random_state` have a leading underscore?

Reply:
> I'm not a huge fan of the names n_outputs_ and keras_expected_n_outputs

`n_outputs_` is what [scikit-learn uses](https://github.com/scikit-learn/scikit-learn/blob/0fb307bf39bbdacd6ed713c00724f8f871d60370/sklearn/neural_network/_multilayer_perceptron.py#L264), and `keras_expected_n_outputs_` is the most descriptive name I could come up with that also matched `n_outputs_`. Happy to discuss renaming the later in a separate issue/PR.

> Why does a certain key, _random_state

Because that's what sklearn does: https://github.com/scikit-learn/scikit-learn/blob/0fb307bf39bbdacd6ed713c00724f8f871d60370/sklearn/neural_network/_multilayer_perceptron.py#L339

Reply:
As per https://github.com/adriangb/scikeras/pull/67/files#r478487342, I renamed `cls_type_` and `input_dtype_`.